File: Top_100_Liked/15-3sum.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 *  Info: 给出一个数组，找出所有不重复的由三个不同数组成的和为0的序列
 *  Key: Two Pointers应用，先排序，然后从0开始遍历，相当于在剩余的
 *       序列里找出和为特定值的两个数，j,k移动的依据是三个数之和大于0还是小于0
 */

// 目前最优解
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        int len = nums.size();
        int i = 0;
        while (i < len-2) {
            int j = i+1, k = len-1;
            while (j < k) {
                if (nums[j]+nums[k]+nums[i] == 0) {
                    vector<int> tmp = {nums[i], nums[j], nums[k]};
                    res.push_back(tmp);
                    j ++, k --;
                    while (j < k && nums[j] == nums[j-1]) j++;
                    while (j < k && nums[k] == nums[k+1]) k--;
                } else if (nums[j]+nums[k]+nums[i] < 0) {
                    j ++;
                } else {
                    k --;
                }
            }
            i ++;
            while (i < len-2 && nums[i]==nums[i-1]) i++;
        }
        return res;
    }
};

File: LeetCode-JS/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

/*
    Two Pointers问题，先排序，再从每个数开始将问题转化为两数之和问题
    关键是排除重复的结果
*/

var threeSum = function(nums) {
    const ans = []
    if (nums.length === 0) return ans

    // 小到大排序
    nums.sort((a, b) => a - b);

    for (let i = 0; i < nums.length-2; i ++ ) {
        if (i > 0 && nums[i] === nums[i-1]) continue;

        let [j, k] = [i+1, nums.length-1];
        while (j < k) {
            const sum = nums[i] + nums[j] + nums[k];
            if (sum === 0) {
                ans.push([nums[i], nums[j], nums[k]]);
                // 排除重复值
                while (j+1 < k-1 && nums[j] === nums[j+1] && nums[k] === nums[k-1]) {
                    j++;
                    k--;
                }
                j++;
                k--;
            } else if (sum < 0) {
                j++;
            } else {
                k--;
            }
        }
    }
    return ans;
};

console.log(threeSum([-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6]))